symengine/sets.cpp

```diff
@@ -464,4 +559,253 @@ RCP<const Boolean> Union::contains(const RCP<const Basic> &o) const
     return boolean(false);
 }
 
+Complement::Complement(const RCP<const Set> &universe,
+                       const RCP<const Set> &container)
+    : universe_(universe), container_(container)
+{
+    SYMENGINE_ASSIGN_TYPEID()
+}
+
+hash_t Complement::__hash__() const
+{
+    hash_t seed = COMPLEMENT;
+    hash_combine<Basic>(seed, *universe_);
+    hash_combine<Basic>(seed, *container_);
+    return seed;
+}
+
+bool Complement::__eq__(const Basic &o) const
+{
+    if (is_a<Complement>(o)) {
+        const Complement &other = down_cast<const Complement &>(o);
+        return unified_eq(universe_, other.universe_)
+               and unified_eq(container_, other.container_);
+    }
+    return false;
+}
+
+int Complement::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Complement>(o))
+    const Complement &other = down_cast<const Complement &>(o);
+    int c1 = unified_compare(universe_, other.universe_);
+    if (c1 != 0)
+        return c1;
+    else
+        return unified_compare(container_, other.container_);
+}
+
+RCP<const Boolean> Complement::contains(const RCP<const Basic> &a) const
+{
+    return logical_and(
+        {universe_->contains(a), logical_not(container_->contains(a))});
+}
+
+RCP<const Set> Complement::set_union(const RCP<const Set> &o) const
+{
+    // A' U C = (A n C')'
+    RCP<const Set> ocomplement = o->set_complement(universe_);
+    RCP<const Set> intersect
+        = SymEngine::set_intersection({container_, ocomplement});
+    return intersect->set_complement(universe_);
+}
+
+RCP<const Set> Complement::set_intersection(const RCP<const Set> &o) const
+{
+    return SymEngine::set_intersection({rcp_from_this_cast<const Set>(), o});
+}
+
+RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
+{
+    auto newuniv = SymEngine::set_union({o, universe_});
+    return container_->set_complement(newuniv);
+}
+
+RCP<const Set> set_union(const set_set &in, bool solve)
+{
+    set_set input;
+    if (solve == false && in.size() > 1)
+        return make_rcp<const Union>(in);
+    set_basic combined_FiniteSet;
+    for (auto it = in.begin(); it != in.end(); ++it) {
+        if (is_a<FiniteSet>(**it)) {
+            const FiniteSet &other = down_cast<const FiniteSet &>(**it);
+            combined_FiniteSet.insert(other.get_container().begin(),
+                                      other.get_container().end());
+        } else if (is_a<UniversalSet>(**it)) {
+            return universalset();
+        } else if (not is_a<EmptySet>(**it)) {
+            input.insert(*it);
+        }
+    }
+    if (input.empty()) {
+        return finiteset(combined_FiniteSet);
+    } else if (input.size() == 1 && combined_FiniteSet.empty()) {
+        return rcp_dynamic_cast<const Set>(*input.begin());
```

Comment:
actually , I only shifted already [implemented](https://github.com/symengine/symengine/pull/1053) `set_union` to cpp file.
Btw, `rcp_dynamic_cast` is not needed here. I will remove them.